### PR TITLE
[ENH] Add scoring direction to ForecastingOptunaSearchCV

### PR DESCRIPTION
**Pull request by gareth-brown-86**
#### What does this implement/fix? Explain your changes.

Currently the optuna study assumes that we are trying to minimise our metric.
We have changed it so it uses the get_tag("lower_is_better") from the passed scorer.
So if the scoring metric should be maximised, the study will  do that.
#### Does your contribution introduce a new dependency? If yes, which one?
No

#### Did you add any tests for the change?
No tests were added for this.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Review by yarnabrina — COMMENTED** [review]
`scoring` is already being validated through `check_scoring`. As far as I can see, you should be able able to directly call `get_tag`, without needing a separate function to check for type and etc.

Can you please tell me why is this not enough?

**Comment by gareth-brown-86** [comment]
> `scoring` is already being validated through `check_scoring`. As far as I can see, you should be able able to directly call `get_tag`, without needing a separate function to check for type and etc.
> 
> Can you please tell me why is this not enough?

@fkiraly  suggested that the use could pass a metric that wasnt a class, and so not have a tag. But I think you make a good point that as it has been run through check_scoring, does that guarantee that it has the tag? 
https://github.com/sktime/sktime/pull/6839#pullrequestreview-2199461391

**Comment by yarnabrina** [comment]
Seeing the docstring, I think callable metrics (I assume this is the only form of non-class metrics which are not already a registered name, may be wrong) must be of lower-is-better type. So, I feel the new function was possibly not necessary, but @fkiraly knows best, with almost encyclopedic knowledge over the entire codebase :)

**Comment by fkiraly** [comment]
In this case, you knew better than me, @yarnabrina!

Indeed you are right, and I forgot about this, `check_scoring` coerces to an `sktime` `BaseMetric`, so it should have the tag after we ran it.

In either case, perhaps we should add a test for the callable case, e.g., in `get_test_params` to make sure it does not break.

**Comment by fkiraly** [comment]
> with almost encyclopedic knowledge over the entire codebase :)

That would be saying too much. It's more of a continuous effort to structure the chaos in my brain.

**Review by fkiraly — CHANGES_REQUESTED** [review]
@yarnabrina is right, we do not need to handle the callable case separately, in terms of tags.

**Comment by gareth-brown-86** [comment]
@yarnabrina @fkiraly So do you suggest I keep it in a simple function, (without the checks I currently added), or just putting the logic inline?
eg inline:

```python
scoring_direction = "minimize" if scoring.get_tag("lower_is_better") else "maximize"
study = optuna.create_study(
    direction=scoring_direction
)
```

**Review by yarnabrina — CHANGES_REQUESTED** [review]
Can we have these please?

1. a new test parameter covering callable scorer
2. can we also cover other arguments for `optuna.create_study`?

Ref. https://optuna.readthedocs.io/en/stable/reference/generated/optuna.create_study.html

**Comment by gareth-brown-86** [comment]
> Can we have these please?
> 
>     1. a new test parameter covering callable scorer
> 
>     2. can we also cover other arguments for `optuna.create_study`?
> 
> 
> Ref. https://optuna.readthedocs.io/en/stable/reference/generated/optuna.create_study.html
1. I don't really know what you mean with this test parameter, would you mind explaining further. At the moment, we have three scoring objects we try, `mean_absolute_percentage_error` and `MeanAbsolutePercentageError` from `sktime.performance_metrics.forecasting` and just a string `"MeanAbsolutePercentageError(symmetric=True)"`. 
2. Regarding the other arguments within the `optuna.create_study`, I would think it makes sense to add them in a separate PR. This is because, if we have a scorer that needs to be optimised, at the moment it would be minimised, so I see that as fixing the current functionality. This is different than the other arguments all add functionality. (note that the argument that changes the sampler, is already in a PR #6823)

**Comment by fkiraly** [comment]
> I don't really know what you mean with this test parameter

Test parameter settings live in the `get_test_params` method of the estimator, see the extension template in `extension_templates`. The function should return a list of dicts, with parameter settings that cover the different types of instances, to ensure all parts of branching code in the estimator are run.

File: sktime/forecasting/model_selection/_tune.py [hunk 1]
```diff
@@ -1913,7 +1913,10 @@ def _run_search(self, y, X, cv, scoring, scoring_name, sampler):
         for (
             param_grid_dict
         ) in self._param_grid:  # Assuming self._param_grid is now a list of dicts
-            study = optuna.create_study(direction="minimize", sampler=sampler)
+            scoring_direction = (
+                "minimize" if scoring.get_tag("lower_is_better") else "maximize"
+            )
+            study = optuna.create_study(direction=scoring_direction, sampler=sampler)
```
Comment (fkiraly) [inline sktime/forecasting/model_selection/_tune.py:1919]: could you kindly check this line? The "from" changed after merging the "scoring direction". I tried to merge sensibly.

Comment (gareth-brown-86) [reply sktime/forecasting/model_selection/_tune.py:1919]: Looks good to me

**Comment by gareth-brown-86** [comment]
> > I don't really know what you mean with this test parameter
> 
> Test parameter settings live in the `get_test_params` method of the estimator, see the extension template in `extension_templates`. The function should return a list of dicts, with parameter settings that cover the different types of instances, to ensure all parts of branching code in the estimator are run.

So in this example, would you want a scorer where `scoring.get_tag("lower_is_better")` is false?

File: sktime/forecasting/model_selection/_tune.py [hunk 2]
```diff
@@ -1814,7 +1814,7 @@ def _fit(self, y, X=None, fh=None):
             ascending=scoring.get_tag("lower_is_better")
         )
         self.cv_results_ = results
-        self.best_index_ = results[f"mean_{scoring_name}"].idxmin()
+        self.best_index_ = results.loc[:, f"rank_{scoring_name}"].argmin()
```
[the remark this answers is not in the document]
Comment (gareth-brown-86) [reply sktime/forecasting/model_selection/_tune.py:1817]: yeah this is correct. The code change here to use the rank and the argmin is consistent with the BaseGridSearch

**Comment by fkiraly** [comment]
I suppose, what @yarnabrina wants is:

* a setting with a scorer that has the tag with opposite sign, yes
* a scorer that is a callable but not an sktime metric
* also, the examples should optimally have some variety, in terms of samplers (added in the other PR), but that is strictly speaking out of scope for this PR

**Comment by yarnabrina** [comment]
Yes, I will prefer to have a coverage of all 6 possibilities arising from following:

1. 3 types of scoring objects (a callable, a class instance and string representation of a class instance based on bove conversation)
2. 2 types of direction (maximisation and minimisation)

Of course with other parameters, e.g. sampler etc. there is more choice but that can be distributed across these examples also. And this should be covered for all tuners if not already covered, but scope of this PR is optuna one only.

(I am not sure if we need to distinguish between a sktime/sklearn metric or a custom function, did not test myself.)

**Comment by fkiraly** [comment]
> Yes, I will prefer to have a coverage of all 6 possibilities arising from following:

I think in line with the "avoid combinatorial explosion" case, perhaps it is better to have subcase coverage but not product case coverage - i.e., all three objects, all two directions but not necessarily products.

So, 4 cases should suffice? Default, plus three cases that amongst them cover the three types of metrics objects, and the two directions.

**Comment by gareth-brown-86** [comment]
@

> > Yes, I will prefer to have a coverage of all 6 possibilities arising from following:
> 
> I think in line with the "avoid combinatorial explosion" rationale for testing complex combination cases, perhaps it is better to have subcase coverage but not product case coverage - i.e., all three objects, all two directions but not necessarily products.
> 
> So, 4 cases should suffice? Default params (dict empty except for mandatory params), plus three cases that amongst them cover the three types of metrics objects, and the two directions.

@yarnabrina @fkiraly 
I have added in the 4th case (change of direction). The other three cases are already covered. 
I have tested using `check_estimator` and all tests passed.

Is there anything else needed to be done (I dont think any changes to the documentation of the function needs to be done, as there is no change in interface, and I think the behavior would be the assumed one by the user)

**Review by fkiraly — APPROVED** [review]
Thanks, looks good to me!

File: sktime/forecasting/model_selection/_tune.py [hunk 3]
```diff
@@ -1900,7 +1900,15 @@ def get_test_params(cls, parameter_set="default"):
             "scoring": "MeanAbsolutePercentageError(symmetric=True)",
             "update_behaviour": "no_update",
         }
-        return [params, params2, params3]
+        scorer_with_lower_is_better_false = MeanAbsolutePercentageError(symmetric=True)
```
Comment (yarnabrina) [inline sktime/forecasting/model_selection/_tune.py:1903]: This is fine for testing, but so counterintuitive!

A different example, say with a function that basically returns negative of smape will be preferable to me.

**Review by yarnabrina — APPROVED** [review]

File: sktime/forecasting/model_selection/_tune.py (hunk 3 above)
Comment (fkiraly) [reply sktime/forecasting/model_selection/_tune.py:1903]: I think there is no class that currently has `lower_is_better` as `False`.

**Pull request merged**